### PR TITLE
Task/NVPE-10: Create NIM Account controller

### PR DESCRIPTION
**Pull request by TomerFi**
## Description

- Added a controller reconciling for NIM accounts.
- Added test cases for the controller including mocks for NVIDIA NIM API.
- Added a "hack" script for troubelshooting NVIDIA API access.

Jira: https://issues.redhat.com/browse/NVPE-10

## How Has This Been Tested?

Tested manully on my station usign the script and helper code.
EDIT (Nov. 12):
Also tested against a live cluster, tests described in the following [google doc](https://docs.google.com/document/d/1GlYxgkyYdIAs0mK2GeuHv3xCpUAQNl34__pxB67P1pw/edit?usp=sharing).

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: controllers/nim_account_controller.go
Comment (lburgazzoli) [inline controllers/nim_account_controller.go:93]: it seems that the resources affected by the cleanup have the `Account` as owner, I guess they get cleaned up automatically by k8s gc, so may not be needed ?

File: controllers/nim_account_controller.go
Comment (lburgazzoli) [inline controllers/nim_account_controller.go:118]: given the update status is updated every time the function return, it seems there is a case where the conditions are reporting wrong information. i.e. in the case the function returns as a consequence of `utils.GetAvailableNimRuntimes()` reporting an error, then the condition would report i.e. "api key secret not available", which may not be true.

not sure if that is expected

File: controllers/nim_account_controller.go
Comment (lburgazzoli) [inline controllers/nim_account_controller.go:238]: not sure if it is applicable, but eventually you can use Server Side Apply instead of `CreateOrUpdate` which may reduce calls and conflicts

File: controllers/nim_account_controller.go
Comment (TomerFi) [reply controllers/nim_account_controller.go]: Resolved in https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/7929b53bb3cbda70594b2aa0169f26d34bf92722.

File: controllers/nim_account_controller.go
Comment (TomerFi) [reply controllers/nim_account_controller.go:119]: Resolved in https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/8f87716a9eec21bf16ce5cf1143cce3ad9295252.

File: main.go
Comment (xieshenzh) [inline main.go:87]: Usually, a controller needs rbac:

```suggestion
// +kubebuilder:rbac:groups=nim.opendatahub.io,resources=accounts,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=nim.opendatahub.io,resources=accounts/status,verbs=get;update;patch
// +kubebuilder:rbac:groups=nim.opendatahub.io,resources=accounts/finalizers,verbs=update
```

Comment (TomerFi) [reply main.go:87]: Is this required?
We'e not creating, patching, or deleting accounts.

File: controllers/nim_account_controller.go
Comment (TomerFi) [reply controllers/nim_account_controller.go]: Resolved in https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/de002501c908763dce09e79acc6551f2ea8e5551.

File: controllers/nim_account_controller.go
Comment (lburgazzoli) [inline controllers/nim_account_controller.go:354]: nit: given the usage path, at this point probably the GVK of the account should be set so it is probably not needed to resolve it via the scheme.

Comment (lburgazzoli) [reply controllers/nim_account_controller.go:354]: note: this is not a blocker, just a suggestion but it seems there are some issues that prevent to implement it as described

Comment (TomerFi) [reply controllers/nim_account_controller.go]: Added a comment referring this in https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/53cf8faa9f79792e0ab12174d53bc67bccc3b3b7.

Basically, what happens is that while working as expected in a live cluster, when running using envtest, the object's TypeMeta is being stripped. See the attached screenshots:

**Live**
image

**EnvTest*
image

Comment (TomerFi) [reply controllers/nim_account_controller.go]: This was tested with the latest envtest (0.19.1) using kube 1.26, 1.29, and 1.30.

File: controllers/nim_account_controller.go
Comment (TomerFi) [reply controllers/nim_account_controller.go]: You know what @lburgazzoli, I'm just reliazing something Yes,  you were right, we didn't need to delete the resources when an acount is being deleted, k8s will handle this.
But the func I removed in the above commit, was also used for when the API key secret is deleted. And this still needs to be handled.

So, I need to bring back the cleanup utility func  for cleanups, but only use it when the secret is deleted.

Comment (TomerFi) [reply controllers/nim_account_controller.go]: (re)Resolved in https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/15dccb2c6f0d1b9814b7a5fe1ba75c8c7a6d4723.

File: controllers/constants/constants.go
Comment (TomerFi) [inline controllers/constants/constants.go:80]: @xieshenzh @etirelli @mpaulgreen

We reconcile a _ConfigMap_, a _Template_, and a pull _Secret_, named _nvidia-nim-data_, _nvidia-nim-serving-template_, and _nvidia-nim-image-pull_ respectively.

If we change the names to something more dynamic, i.e., based on the triggering _Account_ name, please correct me if I'm wrong, but we won't need the admission webhook in #288.

Comment (xieshenzh) [reply controllers/constants/constants.go:80]: I thought the names are dynamic, that's why the names are stored in the status of the Account CR.

The name of the Account is supposed to be provided by the user.

File: controllers/nim_account_controller.go
Comment (terrytangyuan) [inline controllers/nim_account_controller.go:149]: nit: Consider making an util function to log + set condition since these are repeated many times

File: controllers/nim_account_controller.go
Comment (terrytangyuan) [inline controllers/nim_account_controller.go:264]: nit: cmap -> cm

File: controllers/nim_account_controller.go
Comment (terrytangyuan) [inline controllers/nim_account_controller.go:293]: Are there reusable constants?

File: hack/verify_nvidia_nim_api.go
Comment (terrytangyuan) [inline hack/verify_nvidia_nim_api.go:78]: These can be combined to one print?

Comment (TomerFi) [reply hack/verify_nvidia_nim_api.go:78]: Do you mean `fmt.Println(fmt.Sprintf...`?
I can try, but I think this will make _golint-ci_ mad 😁 .

File: controllers/constants/constants.go
Comment (TomerFi) [reply controllers/constants/constants.go:80]: That was the goal, but we were asked to keep these identical to what we have in TP. I think you also commented on this part in the ADR. But maybe I'm wrong.

I'll change these to be dynamic, which will mean that we fully support multiple accounts in the same namespace. If that's the case, why do we need an admission webhook validating that only one exists in a namespace?

Comment (TomerFi) [reply controllers/constants/constants.go:80]: I missed the _status_ perms, fixed here: https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/974306614cbb9c5c1ad61ec5ec10308e2d7dfdab.

We now have:
```
// +kubebuilder:rbac:groups=nim.opendatahub.io,resources=accounts,verbs=get;list;watch;update
// +kubebuilder:rbac:groups=nim.opendatahub.io,resources=accounts/status,verbs=get;list;watch;update
```

I think this should suffice; please resolve if you agree or comment if you do not. 😄

File: controllers/nim_account_controller.go
Comment (TomerFi) [reply controllers/nim_account_controller.go:293]: I can move it to a constant if you prefer.

File: hack/verify_nvidia_nim_api.go
Comment (TomerFi) [reply hack/verify_nvidia_nim_api.go:78]: As expected, `fmt.Println(fmt.Sprintf("Got %d models info successfully", len(models)))`, breaks _golangci-lint_:

image

I can do: `fmt.Printf("Got %d models info successfully\n", len(models))`, but I don't usually prefer to include line breaks in string literals.

Your call.

File: controllers/constants/constants.go
Comment (xieshenzh) [reply controllers/constants/constants.go:80]: > That was the goal, but we were asked to keep these identical to what we have in TP. I think you also commented on this part in the ADR. But maybe I'm wrong.
> 
> I'll change these to be dynamic, which will mean that we fully support multiple accounts in the same namespace. If that's the case, why do we need an admission webhook validating that only one exists in a namespace?

We don't need the wehbook.

I don't think we need the operator at all for NIM enablement alone, if everything is static.
We only need to change the old CronJob to run go code instead of shell scripts, so that it would appear more maintainable.

File: main.go
Comment (TomerFi) [reply main.go:88]: I missed the _status_ perms, fixed here: https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/974306614cbb9c5c1ad61ec5ec10308e2d7dfdab.

We now have:
```
// +kubebuilder:rbac:groups=nim.opendatahub.io,resources=accounts,verbs=get;list;watch;update
// +kubebuilder:rbac:groups=nim.opendatahub.io,resources=accounts/status,verbs=get;list;watch;update
```

I think this should suffice; please resolve if you agree or comment if you do not. 😄

File: controllers/constants/constants.go
Comment (TomerFi) [reply controllers/constants/constants.go:80]: I'm afraid I have to disagree with the job suggestion.

I agree with you that the webhook is unnecessary. This is exactly the point I was trying to make with the dynamic naming convention, but as pointed out in the parallel Slack thread, we need to involve the dashboard people in this.

File: controllers/nim_account_controller.go
Comment (TomerFi) [reply controllers/nim_account_controller.go]: Resolved in https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/b8293f4d86d0821c78a8b52c56a7720f8ed4228f.

File: controllers/nim_account_controller.go
Comment (TomerFi) [reply controllers/nim_account_controller.go:149]: Also, while trying this, I noticed I missed an unknown condition. Resolved here https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/b8293f4d86d0821c78a8b52c56a7720f8ed4228f.

Comment (TomerFi) [reply controllers/nim_account_controller.go:149]: Let me know if you want me to do the utility funcs regardless.

File: main.go
Comment (xieshenzh) [reply main.go:88]: I am resolving this. But please double check the logs of the operator after it runs for a few hours, and make sure there are not issues related to permissions in the logs.

Comment (xieshenzh) [reply main.go:88]: And don't forget to test updating and deleting Account CR.
These events could require additional permissions.

File: controllers/nim_account_controller.go
Comment (israel-hdez) [inline controllers/nim_account_controller.go:184]: What is the motivation for the cleanup in case of errors?

I have the thought that this can break momentarily a working environment during a [controller resync](https://github.com/kubernetes-sigs/controller-runtime/blob/8e29c6387e83b25306d59b87c29424f3e59ca255/designs/cache_options.md#L123).

File: controllers/utils/nim.go
Comment (israel-hdez) [inline controllers/utils/nim.go:97]: Configure the client with timeouts to prevent requests to hang forever....

File: controllers/nim_account_controller.go
Comment (TomerFi) [reply controllers/nim_account_controller.go:72]: Good point!
The Secret is created by ODH Dashboard.

@olavtar @yzhao583 @andrewballantyne
Can you please confirm the Secret encapsulating the API key had the following label:

```
"opendatahub.io/managed": "true"
```

cc @mpaulgreen @xieshenzh

File: controllers/nim_account_controller.go
Comment (TomerFi) [reply controllers/nim_account_controller.go:325]: Resolved in https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/85231d0a2756775c6a6f989c4e992f063188da4e.

File: hack/verify_nvidia_nim_api.go
Comment (TomerFi) [reply hack/verify_nvidia_nim_api.go]: Resolved in https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/85231d0a2756775c6a6f989c4e992f063188da4e.

File: controllers/nim_account_controller.go
Comment (TomerFi) [reply controllers/nim_account_controller.go:184]: We don't clean up for all errors, only for errors relating to the API key validation. The motivation is to remove the model info from the cluster in case of a failed validation. We can remove this if you prefer. I don't think it was a requirement. It may be something that we did in TP and may no longer be required here since we use status conditions.

@mpaulgreen, @xieshenzh, Any thoughts?

Comment (mpaulgreen) [reply controllers/nim_account_controller.go:184]: @xieshenzh wrote the code in the cronob. So, this is the code that deleted the associated resources - https://github.com/opendatahub-io/odh-dashboard/pull/2959/files#diff-654ddbcfb120f3fa423302828fb6b0e72b2a680a025a7a09bc24ccfc4ea2abd9R103-R110. I am not sure why we want to keep the associated configmap and template if api key is invalid? There could be downstream impact too... @yzhao583 @@olavtar any issues if we dont remove the secret, cm and template? But @israel-hdez this is a very common pattern.

File: controllers/nim_account_controller.go
Comment (olavtar) [reply controllers/nim_account_controller.go:72]: @TomerFi I don't see the label in the nvidia-nim-access secret.

Comment (mpaulgreen) [reply controllers/nim_account_controller.go:72]: @olavtar You have to add this label

File: controllers/nim_account_controller.go
Comment (israel-hdez) [reply controllers/nim_account_controller.go:184]: > I am not sure why we want to keep the associated configmap and template if api key is invalid?

If the API Key is invalid, I'm OK with the clean-up. Although, my question is general to the clean-up rather than specific to this `if` block; i.e. there are other cleanups before this one. And, also, I'm thinking mainly about re-syncs.

AFAIK, the controller will do a re-sync (i.e. it will run a reconcile cycle) automatically after a 10-hour period even if nothing has changed in the cluster. As this reconcile is using external services, a re-sync may encounter temporary issues (e.g. API Key validation, or fetching runtimes returning 500 status code, or the request timing out). Such temporary errors will lead to a clean-up, despite everything is still valid.

Since the controller is returning an error, controller-framework would trigger another reconcile quickly which may re-create again all resources if it succeeds. The small window of the deleted resources is my concern. I haven't seen how these resources are used later.

So, if you don't see an issue with the missing resources for a ¿short? window of time, then I'm good with the clean-up.

File: controllers/nim_account_controller.go
Comment (andrewballantyne) [reply controllers/nim_account_controller.go:72]: Yu's PR https://github.com/opendatahub-io/odh-dashboard/pull/3455#discussion_r1838285420 is not currently doing this.

Comment (mpaulgreen) [reply controllers/nim_account_controller.go:72]: Its being added by @olavtar

Comment (olavtar) [reply controllers/nim_account_controller.go:72]: @mpaulgreen Actually Yu is adding it.

File: controllers/constants/constants.go
Comment (TomerFi) [reply controllers/constants/constants.go:80]: For this stage, we're keeping the hardcoded names; further discussion is in [Slack](https://redhat-internal.slack.com/archives/C07CZDA6DE1/p1731435611515029?thread_ts=1731428744.293129&cid=C07CZDA6DE1).

File: controllers/nim_account_controller.go
Comment (TomerFi) [reply controllers/nim_account_controller.go:189]: > If the API Key is invalid, I'm OK with the clean-up. Although, my question is general to the clean-up rather than specific to this if block; i.e. there are other cleanups before this one. And, also, I'm thinking mainly about re-syncs.

We clean up in four scenarios, all of which qualify as a failure to validate the API key:
1. We failed to fetch the Secret encapsulating the API key; see [here](https://github.com/TomerFi/odh-model-controller/blob/nim-account-controller/controllers/nim_account_controller.go#L149).
2. We got the Secret, but it doesn't have the required key; see [here](https://github.com/TomerFi/odh-model-controller/blob/nim-account-controller/controllers/nim_account_controller.go#L162).
3. We failed to fetch NIM's available runtimes; in this case, we don't have runtime info to validate the key with; see [here](https://github.com/TomerFi/odh-model-controller/blob/nim-account-controller/controllers/nim_account_controller.go#L176).
4.  The actual validation fail; see [here](https://github.com/TomerFi/odh-model-controller/blob/nim-account-controller/controllers/nim_account_controller.go#L189).

> AFAIK, the controller will do a re-sync (i.e. it will run a reconcile cycle) automatically after a 10-hour period even if nothing has changed in the cluster. As this reconcile is using external services, a re-sync may encounter temporary issues (e.g. API Key validation, or fetching runtimes returning 500 status code, or the request timing out). Such temporary errors will lead to a clean-up, despite everything is still valid.

Legitimate concern. Perhaps we should reconsider the cleanups.

> Since the controller is returning an error, controller-framework would trigger another reconcile quickly which may re-create again all resources if it succeeds. The small window of the deleted resources is my concern. I haven't seen how these resources are used later.

The controller only returns an error for scenarios 2-4 from the ones described above. If the Secret is not found, no error is returned; see [here](https://github.com/TomerFi/odh-model-controller/blob/nim-account-controller/controllers/nim_account_controller.go#L150).

> So, if you don't see an issue with the missing resources for a ¿short? window of time, then I'm good with the clean-up.

I think we can live with that, but I would be happy if the dashboard people could comment on this, too.
@olavtar @andrewballantyne

File: main.go
Comment (TomerFi) [reply main.go:88]: @xieshenzh You were right, finalzier permissions are also required:
```
cannot set blockOwnerDeletion if an ownerReference refers to a resource you can't set finalizers on
```

File: controllers/nim_account_controller.go
Comment (mpaulgreen) [inline controllers/nim_account_controller.go:428]: AFAIU this function should also cleanup up the api key secret if present from previous TP release. The other resources are something that is managed by operator so it can remain as you feel.

File: main.go
Comment (TomerFi) [reply main.go:88]: @xieshenzh I documented the tests I performed in this [doc](https://docs.google.com/document/d/1GlYxgkyYdIAs0mK2GeuHv3xCpUAQNl34__pxB67P1pw/edit?usp=sharing) (updated the PR body).

File: controllers/nim_account_controller.go
Comment (terrytangyuan) [reply controllers/nim_account_controller.go:299]: it's ok for now

File: controllers/nim_account_controller.go
Comment (israel-hdez) [reply controllers/nim_account_controller.go:77]: I'm now seeing PR opendatahub-io/odh-dashboard#3455 has the label. I'm resolving this conversation.

**Review by israel-hdez — APPROVED** [review]
/lgtm
/hold

**Pull request merged**
